### PR TITLE
App: change cache location

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

make cache dondional

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/lightning/app/source_code/local.py [hunk 1]
```diff
@@ -28,9 +28,13 @@
 class LocalSourceCodeDir:
     """Represents the source code directory and provide the utilities to manage it."""
 
-    cache_location: Path = Path.home() / ".lightning" / "cache" / "repositories"
-
     def __init__(self, path: Path, ignore_functions: Optional[List[_IGNORE_FUNCTION]] = None) -> None:
+        if "LIGHTNING_VSCODE_WORKSPACE" in os.environ:
+            # Don't use home to store the tar ball. This won't play nice with symlinks
+            self.cache_location: Path = Path(os.path.join("/tmp", ".lightning", "cache", "repositories"))
```
Comment (awaelchli) [inline src/lightning/app/source_code/local.py:34]:
```suggestion
            self.cache_location: Path = Path("/tmp", ".lightning", "cache", "repositories")
```
the join is not necessary

File: src/lightning/app/CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Allow customize `gradio` components with lightning colors ([#17054](https://github.com/Lightning-AI/lightning/pull/17054))
 
 
+- Fix `LocalSourceCodeDir` cache_location to not use home in some certain cases ([#17491](https://github.com/Lightning-AI/lightning/pull/17491))
```
Comment (awaelchli) [inline src/lightning/app/CHANGELOG.md:15]: I think this entry is more for the changed section

**Review by awaelchli — APPROVED** [review]

File: src/lightning/app/CHANGELOG.md (hunk 2 above)
Comment (tchaton) [reply src/lightning/app/CHANGELOG.md]: Done !

**Comment by awaelchli** [comment]
@tchaton What is the question? About the failing jobs? They are sometimes flaky

File: pyproject.toml [hunk 3]
```diff
@@ -63,6 +63,7 @@ extend-select = [
 ]
 ignore = [
     "E731",  # Do not assign a lambda expression, use a def
+    "S108",
```
Comment (carmocca) [inline pyproject.toml:66]: Is this intentional? which rule is this?

Comment (tchaton) [reply pyproject.toml:66]: Yes, this is to avoid ruff to complain about the usage of `/tmp`

**Pull request merged**